lib/Basics/ReadWriteLock.cpp

```diff
@@ -27,79 +27,103 @@ using namespace arangodb::basics;
 
 /// @brief locks for writing
 void ReadWriteLock::writeLock() {
-  std::unique_lock<std::mutex> guard(_mut);
-  if (_state == 0) {
-    _state = -1;
+  if (tryWriteLock())
     return;
+
+  // the lock is either hold by another writer or we have active readers
+  // -> announce that we want to write
+  _state.fetch_add(QUEUED_WRITER_INC, std::memory_order_relaxed);
+
+  std::unique_lock<std::mutex> guard(_writer_mutex);
+  while (true)
+  {
+    auto state = _state.load(std::memory_order_relaxed);
+    while ((state & ~QUEUED_WRITER_MASK) == 0)
+    {
+      // try to acquire lock and perform queued writer decrement in one step
+      if (_state.compare_exchange_strong(state, (state - QUEUED_WRITER_INC) | WRITE_LOCK, std::memory_order_acquire))
```

Comment:
wouldn't `compare_exchange_weak` be better here since it's in a loop ?

Reply:
Yes, in this case a compare_exchange_weak would be sufficient.